### DISCUSSION TITLE
Is there really no way to tell when a form submission is complete?

### DISCUSSION THREAD

**Discussion by dan-cooke** [vercel/next.js#58477] [App Router] [answered]
Hey all,

So we have the following experimental module from react for `useFormStatus` in Next13

```.tsx

declare module '.' {
    interface FormStatusNotPending {
        pending: false;
        data: null;
        method: null;
        action: null;
    }

    interface FormStatusPending {
        pending: true;
        data: FormData;
        method: string;
        action: string | ((formData: FormData) => void | Promise<void>);
    }

    type FormStatus = FormStatusPending | FormStatusNotPending;

    function experimental_useFormStatus(): FormStatus;
}
```

I do not see any way to determine if a form has been submitted successfully? 

I can't seem to find anything about htis in the docs - I can only see how to do optimistic repsonse, error handling etc. 

Do we really have to do some hack with `useEffect` and determine the form has went through `pending: true` to `pending: false` again?

I think the idea is that we are supposed to redirect from our actions once successful, but its very inconenvient sometimes - you don't want to duplicate an entire UI just to display a success toast or something

**Comment by icyJoseph** [comment]
https://react.dev/reference/react-dom/hooks/useFormStatus

**Reply by dan-cooke** [reply]
That is literally the exact hook i referred to in my post, it does not allow you to determine when a form has been submitted - just "submitting"

**Reply by icyJoseph** [reply, marked as answer]
Then I guess you want https://react.dev/reference/react-dom/hooks/useFormState ? Which can use the result of the action call, which will let you know if the form has been submitted.

**Reply by dan-cooke** [reply]
Thank you not sure how I missed this, I don’t see it mentioned in the next documentation
